### ISSUE TITLE
Figure out how to keep Typescript definitions of core objects in sync.

### ISSUE THREAD

**Issue by bk3c** [bloom-housing/bloom#283]
Spun off from #227 now that we've decided to go ahead with TypeORM. The ideal state is that we have exactly one place for definitions of key objects (e.g. Listing, Address, Application) that can then be consumed by both front-end apps and back-end services. These are currently in `shared/core`, but as we start to have decorated TypeORM definitions within each service, that's something new that we'll need to keep in sync, so it would be good to figure out the most-DRY code structure.

If it's not feasible to have things in exactly one place, I think it's probably reasonable to have things in two places with some automated tests to keep them in sync. If we go that route, I'd love to see some template tests as part of this issue, which we can then extend as we continue to build out the objects.

Another pattern that seems reasonable to consider is that object definitions can live with the service that's the storage of record for that object, and then the typescript definition can be consumed by the other services and frontend apps as needed. I was initially hesitant to go down this road, since we might have some services customized in a language other than TS on Node for specific implementations, but thinking about it more I think it's just as easy for those customized services to just maintain and statically export TS interfaces no matter what language or stack the underlying code is written in. And reference implementations of services are still intended to be all Typescript for the foreseeable future.

**Comment by bk3c** [comment]
A few links I found with how people have approach this with TypeORM in the past:

- [~2018 TypeORM Github issue](https://github.com/typeorm/typeorm/issues/2841), which then points to some other issues with potential solutions.
- [Short /r/Typescript thread](https://www.reddit.com/r/typescript/comments/b9gz4c/best_practices_for_sharing_types_between_frontend/) which mentions the TypeORM shim that I mentioned in standup.
- [Longer /r/reactjs thread](https://www.reddit.com/r/reactjs/comments/ey282l/fullstack_web_app_in_react_typescript_graphql/) which focuses more on codegen solutions from GraphQL. I'm still open to this approach on at least the front-end, but I wonder if GraphQL would be cumbersome for service->service communication (probably no worse than REST?)
- The NX team seems [to have baked this in completely](https://nx.dev/angular/tutorial/07-share-code) to their system at this point. I'm not sure how mature the React tooling is compared to Angular, and we're certainly not switching that at this point, but maybe it will provide some inspiration?

**Comment by jaredcwhite** [comment]
@bk3c Thanks for the background research on this. I'll dig into it further! I think since the Core package and the other services are all in the same repo there must some  way to automate the interface generation to stay synchronized with the underlying data store, so will try to get a handle on that.

**Comment by jaredcwhite** [comment]
So… I don't know for sure if this works because I haven't booted up the new listing service/database yet, but I was able to work through all the Typescript errors in order to get Next.js to compile. Here's the branch:

https://github.com/bloom-housing/bloom/tree/peace-between-typeorm-and-nextjs

Basically some updates to the Babel/TS config and having to explicitly declare field types in the TypeORM decorators. Not sure why that's necessary rather than TypeORM interfering from the TS types themselves, but it's not a huge deal IMHO. At least the "duplication" is all in one code block so we don't have to go anywhere else to look at type information.

The point is that with this approach the public apps can simply import interfaces from `@bloom-housing/core` and the fact that core is merely re-exporting interfaces from `@bloom-housing/listing-service` is an underlying implementation detail. (We might have to change core to require `@bloom-housing/listing-service` as a dev dependency just so it can access the interfaces, but that seems OK to me. @bk3c thoughts?)

@software-project Can you give this a try with your local database and see if what I did works at all or if I broke the backend? 😂

**Comment by pbn4** [comment]
> A few links I found with how people have approach this with TypeORM in the past:
> 
> * [~2018 TypeORM Github issue](https://github.com/typeorm/typeorm/issues/2841), which then points to some other issues with potential solutions.
> * [Short /r/Typescript thread](https://www.reddit.com/r/typescript/comments/b9gz4c/best_practices_for_sharing_types_between_frontend/) which mentions the TypeORM shim that I mentioned in standup.
> * [Longer /r/reactjs thread](https://www.reddit.com/r/reactjs/comments/ey282l/fullstack_web_app_in_react_typescript_graphql/) which focuses more on codegen solutions from GraphQL. I'm still open to this approach on at least the front-end, but I wonder if GraphQL would be cumbersome for service->service communication (probably no worse than REST?)
> * The NX team seems [to have baked this in completely](https://nx.dev/angular/tutorial/07-share-code) to their system at this point. I'm not sure how mature the React tooling is compared to Angular, and we're certainly not switching that at this point, but maybe it will provide some inspiration?

@bk3c At first I understood that models should be reusable between various backend services, not with the frontend. Isn't swagger plugin for NestJS solving exactly this problem?

**Comment by software-project** [comment]
@pbn4 and had another look at typescript today. Michal made some fixes for workspaces settings. We have noticed that there were some conflicts with master updates, so we have fixed them (mostly Michał). @jaredcwhite similar issues occurred when I pulled your branch, so I wasn't easily able to test it. 
Regarding the approach, maybe I'm missing something here, but why don't we just do:
`interface IListing`
`export class Listing implements IListing`
I believe thats the core idea behind interfaces, to give a scaffold for a class. Sure there is a little bit of duplication, but still thats how interfaces are supposed to be working. We can use the interface on the frontend, no shims or other stuff. Also I don't like EntitySchema syntax, it looks like huge json config thing.
